### PR TITLE
Reference read with sort #129

### PR DESCRIPTION
**Pull request by shinyichen**
This pull request cover issue #129 and relevant test cases.

* Parse sort from URI, also implement multiple sort columns.
* Sort on Reference.read(): 1) Use sort if available, else used row_order from table-display, otherwise don't specify sorting.
* sort() method that makes a new Reference with a different sort.

### CODE REVIEW

File: test/specs/reference/tests/02.reference_sort.js [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+        var multipleEntityUri = options.url + "/catalog/" + catalog_id + "/entity/" + schemaName + ":"
+            + tableName;
+
+
+        // Test Cases:
+        describe('Reference.sort() method, ', function() {
+            var reference, reference2, page;
+            var limit = 10;
+
+            it('resolve should return a Reference object that is defined.', function(done) {
```
Comment (RFSH) [inline test/specs/reference/tests/02.reference_sort.js]: This test case should not be here. If you want to access the reference object, just use `beforeAll`.
```javascript
beforeAll(function(done) {
  options.ermRest.resolve(singleEnitityUri, {cid: "test"}).then(function(response) {
    reference = response.contextualize.record;
    done();
  }, function(err) {
    console.dir(err);
    done.fail();
  })
});
```

**Comment by jrchudy** [comment]
The changes to ermrestJS look good to me.

File: js/reference.js [hunk 2]
```diff
@@ -396,6 +446,29 @@ var ERMrest = (function(module) {
...
+        sort: function(sort) {
+
+            // make a Reference copy
+            var newReference = _referenceCopy(this);
+            // change ._sort
+            if (!sort || sort.length === 0) {
+                delete newReference._sort;
+            }
+            else {
+                verify((sort instanceof Array), "input should be an array");
```
Comment (robes) [inline js/reference.js]: You could also add a deeper verification here.

```javascript
function isValidElement (element, index, array) {
    return (typeof element == 'object' &&
            typeof element.column == 'string' &&
            typeof element.order == 'string');
}
verify(sort.every(isValidElement), "invalid arguments in array");
```

MDN says that Array.every is supported fairly well.
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/every

**Comment by karlcz** [comment]
In order to have `@before(...)` or `@after(...)` paging modifiers on References, you're going to have to have a sort on the Reference. So, I think the behavior described in this issue is not quite right. We should set a sort if it is missing and also augment any specified sort to include a primary key (if not already there) so that any subsequent Page result from Reference.read() will be able to expose its next/previous page references...

**Comment by shinyichen** [comment]
How about keeping a shortest key in Reference and use that when sort is missing? And I will add that in the Paging issue #130

**Comment by karlcz** [comment]
You also need to append the shortest key to the sort if the requested
sort doesn't include at least one complete key.

**Comment by shinyichen** [comment]
Yes, I already implemented that part here.

**Pull request merged**